src/Region.cpp

```diff
@@ -0,0 +1,808 @@
+  double min_dist = INFINITY;
+
+  /* Find the minimum distance to one of the Region's nodes */
+  std::vector<Region*>::iterator iter;
+  for (iter = _nodes.begin(); iter != _nodes.end(); ++iter) {
+    curr_dist = (*iter)->minSurfaceDist(point, azim, polar);
+
+    /* If the distance to Cell is less than current min distance, update */
+    if (curr_dist < min_dist)
+      min_dist = curr_dist;
+  }
+  return min_dist;
+}
+
+
+/**
+ * @brief Create a duplicate of the Region.
+ * @return a pointer to the clone
+ */
+Region* Region::clone() {
+
+  /* Instantiate appropriate class type for the clone */
+  Region* clone;
+  if (dynamic_cast<Intersection*>(this))
+    clone = new Intersection();
+  else if (dynamic_cast<Union*>(this))
+    clone = new Union();
+  else if (dynamic_cast<Complement*>(this))
+    clone = new Complement();
+  //else if (dynamic_cast<Halfspace*>(this))
+  //  clone = new Halfspace(dynamic_cast<Halfspace*>(this)->getHalfspace(), 
+  //                        dynamic_cast<Halfspace*>(this)->getSurface());
+
+  /* Add this region's nodes to the cloned region */
+  std::vector<Region*>::iterator iter;
+  for (iter = _nodes.begin(); iter != _nodes.end(); iter++)
+    clone->addNode((*iter));
+  return clone;
+}
+
+
+/**
+ * @brief Constructor sets the type of Region (INTERSECTION).
+ */
+Intersection::Intersection() {
+  _region_type = INTERSECTION;
+}
+
+
+/**
+ * @brief Determines whether a Point is contained inside the Intersection.
+ * @details Queries each of the Intersection's nodes to determine if the
+ *          Point is within the Intersection. This point is only inside the
+ *          Intersection if it is contained by each and every node.
+ * @param point a pointer to a Point
+ * @returns true if the Point is inside the Intersection; otherwise false
+ */
+bool Intersection::containsPoint(Point* point) {
+
+  /* Query each of the Intersection's nodes */
+  std::vector<Region*>::iterator iter;
+  for (iter = _nodes.begin(); iter != _nodes.end(); iter++) {
+    if (!(*iter)->containsPoint(point))
+      return false;
+  }
+  return true;
+}
+
+
+/**
+ * @brief Constructor sets the type of Region (UNION).
+ */
+Union::Union() {
+  _region_type = UNION;
+}
+
+
+/**
+ * @brief Determines whether a Point is contained inside the Union.
+ * @details Queries each of the Union's nodes to determine if the Point
+ *          is within the Union. This point is only inside the
+ *          Union if it is contained by at least one node.
+ * @returns true if the Point is inside the Union; otherwise false
+ */
+bool Union::containsPoint(Point* point) {
+
+  /* Query each of the Intersection's nodes */
+  std::vector<Region*>::iterator iter;
+  for (iter = _nodes.begin(); iter != _nodes.end(); iter++) {
+    if ((*iter)->containsPoint(point))
+      return true;
+  }
+  return false;
+}
+
+
+/**
+ * @brief Constructor sets the type of Region (COMPLEMENT).
+ */
+Complement::Complement() {
+  _region_type = COMPLEMENT;
+}
+
+
```

Comment:
Complement is a unary operator. Better to limit the size of _nodes only one. Also, an extra addNode function should be defined.

Reply:
Will do.